crates/adapters/src/controller/stats.rs

```diff
@@ -199,6 +204,138 @@ pub struct ControllerStatus {
     /// Output endpoint configs and metrics.
     #[serde(serialize_with = "serialize_outputs")]
     outputs: OutputsStatus,
+
+    /// Metrics.
+    pub metrics: Mutex<Vec<ControllerMetric>>,
+}
+
+/// Controller metrics.
+pub struct ControllerMetric {
+    /// Is this a counter, gauge, or histogram?
+    ///
+    /// A counter counts some kind of event. It only goes up.
+    ///
+    /// A gauge reports the most recent value of some measurement. It may
+    /// increase and decrease over time.
+    ///
+    /// A histogram reports a sequence of measured values. Each value of the
+    /// histogram is reported only once, so subsequent reads of the metric will
+    /// not include previously seen values, and if the metric is read twice in
+    /// quick succession the second read is likely to report an empty histogram.
+    kind: MetricKind,
+
+    /// Metric name.
+    key: KeyName,
+
+    /// Optional key-value pairs that provide additional metadata about this
+    /// metric.
+    labels: Vec<(MetricString, MetricString)>,
+
+    /// Unit of measure for this metric, if any.
+    unit: Option<MetricUnit>,
+
+    /// Optional natural language description of the metric.
+    description: Option<MetricString>,
+
+    /// The metric's value.
+    value: MetricValue,
+}
+
+impl
+    From<(
+        CompositeKey,
+        Option<MetricUnit>,
+        Option<MetricString>,
+        MetricValue,
+    )> for ControllerMetric
+{
+    fn from(
+        (composite_key, unit, description, value): (
+            CompositeKey,
+            Option<MetricUnit>,
+            Option<MetricString>,
+            MetricValue,
+        ),
+    ) -> Self {
+        let (kind, key) = composite_key.into_parts();
+        let (name, labels) = key.into_parts();
+        Self {
+            kind,
+            key: name,
+            labels: labels.into_iter().map(|label| label.into_parts()).collect(),
+            unit,
+            description,
+            value,
+        }
+    }
+}
+
+impl Serialize for ControllerMetric {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: Serializer,
+    {
+        let len = 3
```

Comment:
This is `3` because the struct has 3 non-empty fields?

Reply:
Yes. I can add a comment. I don't know why serde cares but it does.